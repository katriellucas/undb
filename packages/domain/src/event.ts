import { v4 } from 'uuid'

export interface IEvent<TPayload extends object = object, TMeta = any> {
  id: string
  name: string
  operatorId: string
  payload: TPayload
  timestamp: Date
  meta: TMeta
}

export interface IEventJSON<TPayload extends object = object, TMeta = any> {
  id: string
  name: string
  operatorId: string
  payload: TPayload
  timestamp: string
  meta: TMeta
}

export abstract class BaseEvent<
  TPayload extends object = object,
  TName extends string = string,
  TMeta extends object | undefined = undefined,
> implements IEvent<TPayload>
{
  abstract name: TName
  constructor(
    public readonly payload: TPayload,
    public readonly operatorId: string,
    public readonly meta: TMeta,
    public readonly id = v4(),
    public readonly timestamp = new Date(),
  ) {}

  toJSON(): IEventJSON<TPayload, TMeta> {
    return {
      id: this.id,
      name: this.name,
      operatorId: this.operatorId,
      timestamp: this.timestamp.toISOString(),
      payload: this.payload,
      meta: this.meta,
    }
  }
}
