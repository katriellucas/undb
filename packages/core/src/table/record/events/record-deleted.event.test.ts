import { createTestTable } from '../../fixtures'
import { createTestRecord } from '../fixtures'
import { RecordDeletedEvent } from './record-deleted.event'

describe('delete record', () => {
  test('delete record', () => {
    const table = createTestTable()
    const record = createTestRecord()
    const deleteEvent = RecordDeletedEvent.from(table, 'user1', record)
    const snapshotData = {
      payload: deleteEvent.payload,
      operatorId: deleteEvent.operatorId,
      name: deleteEvent.name,
      meta: {
        record: {
          id: deleteEvent.meta.record.id,
          createdBy: deleteEvent.meta.record.createdBy,
          createdByProfile: deleteEvent.meta.record.createdByProfile,
          updatedBy: deleteEvent.meta.record.updatedBy,
          updatedByProfile: deleteEvent.meta.record.updatedByProfile,
          autoIncrement: deleteEvent.meta.record.autoIncrement,
          tableId: deleteEvent.meta.record.tableId,
          values: deleteEvent.meta.record.values,
          displayValues: deleteEvent.meta.record.displayValues,
        },
      },
    }
    expect(snapshotData).toMatchSnapshot()
  })
})
