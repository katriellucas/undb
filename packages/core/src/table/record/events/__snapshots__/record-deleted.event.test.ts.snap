// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`delete record > delete record 1`] = `
{
  "meta": {
    "record": {
      "autoIncrement": undefined,
      "createdBy": undefined,
      "createdByProfile": null,
      "displayValues": {},
      "id": "record",
      "tableId": "tableId",
      "updatedBy": undefined,
      "updatedByProfile": null,
      "values": {},
    },
  },
  "name": "record.deleted",
  "operatorId": "user1",
  "payload": {
    "id": "record",
    "name": "",
    "tableId": "tableId",
    "tableName": "name",
  },
}
`;
